File: Assign1/Go/Src/Concurrency/readerwriter.go
package main

import (
	"sync"
	"time"
	"fmt"
)

func sleep() {
	time.Sleep(1 * time.Millisecond)
}

func reader(m *sync.RWMutex, id int, wg *sync.WaitGroup) {
	defer wg.Done()
	fmt.Println("reader", id, " is ready to read");
	sleep()
	m.RLock()
	fmt.Println("reading", id, " is reading");
	m.RUnlock()
}
func writer(m *sync.RWMutex, id int, wg *sync.WaitGroup) {
	defer wg.Done()
	fmt.Println("writer", id, " is ready to write");
	sleep()
	m.Lock()
	fmt.Println("writer", id, " is writing");
	m.Unlock()

}
func main() {
	t1 := time.Now()

	var m sync.RWMutex
	var wg sync.WaitGroup
	for i := 0; i < 1000; i++ {
		wg.Add(1)
		go reader(&m, i, &wg)
	}
	for i := 0; i < 500; i++ {
		wg.Add(1)
		go writer(&m, i, &wg)
	}

	wg.Wait()

	elapsed := time.Since(t1)
	fmt.Println("All threads elapsed: ", elapsed)

}

File: Assign1/Java/Concurrency/src/ReaderWriter.java


import java.util.Vector;
import java.util.concurrent.locks.ReentrantReadWriteLock;

public class ReaderWriter {
	private ReentrantReadWriteLock lock = new ReentrantReadWriteLock();   

    public static void main(String[] args) {    
    	
    	long startTime=System.currentTimeMillis();  
    	Vector<Thread> threads = new Vector<Thread>();
    	
    	ReaderWriter rw = new ReaderWriter();

    	for(int i = 0; i < 500; i++)
        {

        	Thread iThread =  new Thread(rw.new Writer(i));
        	threads.add(iThread);    
        	iThread.start();
        		
        }
        for(int i = 0; i < 1000; i++)
        {
	       	Thread iThread =  new Thread(rw.new Reader(i));
        	threads.add(iThread);    
        	iThread.start();
        }
        
        for (Thread iThread : threads) {
            try {
              iThread.join();
            } catch (InterruptedException e) {
              e.printStackTrace();
            }
        }
     
        
        long endTime=System.currentTimeMillis();       
        System.out.println("\"All threads elapsed: "+(endTime-startTime)+"ms");
    }
  
    class Writer implements Runnable {
    	private int  id;    	
        public Writer(int id)
        {
            this.id = id;
        }
        public void run()  
        {       	
        	try {
                Thread.sleep(1);
            } catch (InterruptedException e1) {
                e1.printStackTrace();
            } 
        	System.out.println("writer " + id + " is ready to write"); 
        	lock.writeLock().lock();
	        System.out.println("writer " + id + " is writing"); 
	        lock.writeLock().unlock();  
        }  
    }
  
    class Reader implements Runnable {   
    	private int  id;
    	
        public Reader(int id)
        {
            this.id = id;
        }

    	public void run()  
        {   
    		try {
                Thread.sleep(1);
            } catch (InterruptedException e1) {
                e1.printStackTrace();
            }
    		System.out.println("reader " + id + " is ready to read"); 

    		lock.readLock().lock();
            System.out.println("reader " + id + " is reading"); 
            lock.readLock().unlock();
        } 
    }
}

